weld/sir/optimizations/fold_constants.rs

```diff
@@ -0,0 +1,193 @@
+                AssignLiteral(ref lit) => {
+                    values.insert(statement.output.clone().unwrap(), *lit);
+                    true
+                },
+                // Aliases
+                Assign(ref sym) => {
+                    // contains_key instead of get to avoid double mutable borrow
+                    if values.contains_key(sym) {
+                        let value = *values.get(sym).unwrap();
+                        values.insert(statement.output.clone().unwrap(), value);
+                        true
+                    } else {
+                        false 
+                    }
+                }
+                BinOp { ref op, ref left, ref right } if values.contains_key(left) && values.contains_key(right) => {
+                    let left_val = *values.get(left).unwrap();
+                    let right_val = *values.get(right).unwrap();
+                    // If this throws an error, it just means that we don't support evaluating the
+                    // expression right now.
+                    if let Ok(result) = evaluate_binop(*op, left_val, right_val) {
+                        values.insert(statement.output.clone().unwrap(), result);
+                        true
+                    } else {
+                        false
+                    }
+                }
+                // Unsupported for now
+                _ => false
+            };
+
+            if replaced {
+                // If this is true, we computed a constant value for this statement. Replace it
+                // with that constant value.
+                let output_sym = statement.output.clone().unwrap();
+                let kind = AssignLiteral(*values.get(&output_sym).unwrap());
+                let new_statement = Statement::new(Some(output_sym), kind);
+                *statement = new_statement;
+            } else {
+                // We didn't replace this statement with an AssignLiteral, so record the symbols
+                // which it uses.
+                used_symbols.extend(statement.kind.children().cloned());
+            }
+        }
+    }
+
+    // Delete unused symbols. We'll only delete the locals, since deleting the parameters will
+    // require modifying global state.
+    for block in func.blocks.iter_mut() {
+        block.statements.retain(|ref s| {
+            // Keep the statement that defines the symbol if it is used on the RHS of some
+            // expression, and remove it otherwise.
+            if let Some(ref sym) = s.output {
+                used_symbols.contains(sym) || global_params.contains(sym)
+            } else {
+                // If there's no output symbol (i.e., it's a `Merge`), retain it.
+                true
+            }
+        });
+    }
+
+    // TODO(shoumik): is there a better way of doing this...
+    let mut locals = BTreeMap::new();
+    for (k, v) in func.locals.iter() {
+        if used_symbols.contains(k) || global_params.contains(k) {
```

Comment:
`global_params` may contain locals because you added terminator variables to it as well?

Reply:
Yup!